storm-client/test/jvm/org/apache/storm/utils/UtilsTest.java

```diff
@@ -237,4 +247,167 @@ public void checkVersionInfo() {
+                testDescription = "dummy test";
+                topology = null;
+                expectedCycles = 0;
+            }
+
+            CycleDetectionScenario(String testName, String testDescription, StormTopology topology, int expectedCycles) {
+                this.testName = testName.replace(' ', '-');
+                this.testDescription = testDescription;
+                this.topology = topology;
+                this.expectedCycles = expectedCycles;
+            }
+
+            public List<CycleDetectionScenario> createTestScenarios() {
+                List<CycleDetectionScenario> ret = new ArrayList<>();
+                int testNo = 0;
+                CycleDetectionScenario s;
+                TopologyBuilder tb;
+
+                // Base case
+                {
+                    testNo++;
+                    tb = new TopologyBuilder();
+                    tb.setSpout("spout1", new TestWordSpout(), 10);
+                    tb.setBolt("bolt1", new TestWordCounter(), 10).shuffleGrouping("spout1");
+                    tb.setBolt("bolt2", new TestWordCounter(), 10).shuffleGrouping("spout1");
+                    tb.setBolt("bolt11", new TestWordCounter(), 10).shuffleGrouping("bolt1");
+                    tb.setBolt("bolt12", new TestWordCounter(), 10).shuffleGrouping("bolt1");
+                    tb.setBolt("bolt21", new TestWordCounter(), 10).shuffleGrouping("bolt2");
+                    tb.setBolt("bolt22", new TestWordCounter(), 10).shuffleGrouping("bolt2");
+                    s = new CycleDetectionScenario(String.format("(%d) Base", testNo),
+                            "Three level component hierarchy with no loops",
+                            tb.createTopology(),
+                            0);
+                    ret.add(s);
+                }
+
+                // single loop with one bolt
+                {
+                    testNo++;
+                    tb = new TopologyBuilder();
+                    tb.setSpout("spout1", new TestWordSpout(), 10);
+                    tb.setBolt("bolt1", new TestWordCounter(), 10).shuffleGrouping("spout1");
+                    tb.setBolt("bolt2", new TestWordCounter(), 10).shuffleGrouping("spout1");
+                    tb.setBolt("bolt11", new TestWordCounter(), 10).shuffleGrouping("bolt1");
+                    tb.setBolt("bolt12", new TestWordCounter(), 10).shuffleGrouping("bolt1");
+                    tb.setBolt("bolt21", new TestWordCounter(), 10).shuffleGrouping("bolt2");
+                    tb.setBolt("bolt22", new TestWordCounter(), 10).shuffleGrouping("bolt2");
+                    // loop bolt 3  (also connect bolt3 to spout 1)
+                    tb.setBolt("bolt3", new TestWordCounter(), 10).shuffleGrouping("spout1").shuffleGrouping("bolt3");
+                    ret.add(new CycleDetectionScenario(String.format("(%d) One Loop", testNo),
+                            "Four level component hierarchy with 1 cycle in bolt3",
```

Comment:
Should this be "Three Level" component hierarchy?

Reply:
Yes. Fixed.